src/drivers/intel/cavs/hda-dma.c

```diff
@@ -802,146 +723,156 @@ static int hda_dma_pm_context_store(struct dma *dma)
 	return 0;
 }
 
-static int hda_dma_set_cb(struct dma *dma, unsigned int channel, int type,
+static int hda_dma_set_cb(struct dma_chan_data *channel, int type,
 	void (*cb)(void *data, uint32_t type, struct dma_cb_data *next),
 	void *data)
 {
-	struct dma_pdata *p = dma_get_drvdata(dma);
 	uint32_t flags;
 
-	if (channel >= HDA_DMA_MAX_CHANS) {
-		trace_hddma_error("hda-dmac: %d invalid channel %d",
-				  dma->plat_data.id, channel);
-		return -EINVAL;
-	}
-
 	irq_local_disable(flags);
-	p->chan[channel].cb = cb;
-	p->chan[channel].cb_data = data;
-	p->chan[channel].cb_type = type;
+	channel->cb = cb;
+	channel->cb_data = data;
+	channel->cb_type = type;
 	irq_local_enable(flags);
 
 	return 0;
 }
 
 static int hda_dma_probe(struct dma *dma)
 {
-	struct dma_pdata *hda_pdata;
+	struct hda_chan_data *hda_chan;
 	int i;
-	struct hda_chan_data *chan;
+	struct dma_chan_data *chan;
 
 	trace_hddma("hda-dmac :%d -> probe", dma->plat_data.id);
 
-	if (dma_get_drvdata(dma))
+	if (dma->chan)
 		return -EEXIST; /* already created */
 
-	/* allocate private data */
-	hda_pdata = rzalloc(RZONE_SYS_RUNTIME | RZONE_FLAG_UNCACHED,
-			    SOF_MEM_CAPS_RAM, sizeof(*hda_pdata));
-	if (!hda_pdata) {
-		trace_hddma_error("hda-dmac: %d alloc failed",
+	dma->chan = rzalloc(RZONE_SYS_RUNTIME | RZONE_FLAG_UNCACHED,
+			    SOF_MEM_CAPS_RAM, sizeof(struct dma_chan_data) *
+			    dma->plat_data.channels);
+
+	if (!dma->chan) {
+		trace_hddma_error("hda-dmac: %d channels alloc failed",
 				  dma->plat_data.id);
-		return -ENOMEM;
+		goto out;
 	}
-	dma_set_drvdata(dma, hda_pdata);
 
 	/* init channel status */
-	chan = hda_pdata->chan;
+	chan = dma->chan;
 
-	for (i = 0; i < HDA_DMA_MAX_CHANS; i++, chan++) {
+	for (i = 0; i < dma->plat_data.channels; i++, chan++) {
 		chan->dma = dma;
 		chan->index = i;
 		chan->status = COMP_STATE_INIT;
+
+		hda_chan = rzalloc(RZONE_SYS_RUNTIME | RZONE_FLAG_UNCACHED,
+				   SOF_MEM_CAPS_RAM,
+				   sizeof(struct hda_chan_data));
```

Comment:
Could someone actually do some tests? Namely add a print if this allocation fails (might be this the cause of CI failures?)

Reply:
@xiulipan Would any trace_error calls be visible in the SOFCI tests on HDA platforms?

Reply:
@paulstelian97 CI will not find any trace_error if there is no error happen.
But we can run some test.

BTW: are you asking the error logs for this PR or for code base?

Reply:
@xiulipan I would have suggested to add a trace_error temporarily that triggers if rzalloc failed, as I suspect it may be at the root of the issue with all HDA failures.